### PR TITLE
Openssl 1.1

### PR DESCRIPTION
**Pull request by andred**
I can't push to #362, so this is an updated version.

Compared to #362, this fixes a few issues detected so far:
- code involving EVP_CIPHER_CTX_copy() has been fixed
- EVP_CIPHER_CTX_init() / EVP_CIPHER_CTX_reset() / OpenSSLCipher::Reset() usages have been simplified in the OpenSSL 1.1.x case
- class Sha256HMACOpenSSLImpl allocates OpenSSL objects in its constructor now as requested
- this supports LibreSSL again

### CODE REVIEW

File: aws-cpp-sdk-core/include/aws/core/utils/crypto/openssl/CryptoImpl.h [hunk 1]
```diff
@@ -21,10 +21,16 @@
...
 #include <openssl/ossl_typ.h>
 #include <openssl/evp.h>
+#include <openssl/hmac.h>
 #include <openssl/rand.h>
 #include <atomic>
 #include <mutex>
 
+#if defined(LIBRESSL_VERSION_NUMBER) && (OPENSSL_VERSION_NUMBER == 0x20000000L)
+#undef OPENSSL_VERSION_NUMBER
+#define OPENSSL_VERSION_NUMBER 0x1000107fL
```
Comment (singku) [inline aws-cpp-sdk-core/include/aws/core/utils/crypto/openssl/CryptoImpl.h]: How is this version number comes?

File: aws-cpp-sdk-core/include/aws/core/utils/crypto/openssl/CryptoImpl.h [hunk 2]
```diff
@@ -191,7 +202,10 @@ namespace Aws
 
                 virtual size_t GetKeyLengthBits() const = 0;
 
-                EVP_CIPHER_CTX m_ctx;
+                EVP_CIPHER_CTX *m_ctx;
+#if OPENSSL_VERSION_NUMBER < 0x10100003L
```
Comment (singku) [inline aws-cpp-sdk-core/include/aws/core/utils/crypto/openssl/CryptoImpl.h]: This is not good to appear in multiple locations.

File: aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp [hunk 3]
```diff
@@ -229,19 +259,17 @@ namespace Aws
...
             OpenSSLCipher::OpenSSLCipher(OpenSSLCipher&& toMove) : SymmetricCipher(std::move(toMove)),
-                                                                   m_encDecInitialized(false)
+                                                                   m_ctx(nullptr), m_encDecInitialized(false)
             {
-                m_ctx = toMove.m_ctx;
-                toMove.m_ctx.cipher = nullptr;
-                toMove.m_ctx.cipher_data = nullptr;
-                toMove.m_ctx.engine = nullptr;
+                Init();
+                EVP_CIPHER_CTX_copy (m_ctx, toMove.m_ctx);
```
Comment (singku) [inline aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp]: In original impl, there are several fields set up to nullptr, and I checked the code of EVP_CIPHER_CTX_copy, it doesn't cleanup in arg, I guess we should call EVP_CIPHER_CTX_cleanup(toMove.m_ctx) here.

**Comment by singku** [comment]
Why don't we just switch to pointer impl if OPENSSL made the type opaque?

File: aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp [hunk 4]
```diff
@@ -411,14 +454,10 @@ namespace Aws
...
-                {
-                    EVP_CIPHER_CTX_cleanup(&m_ctx);
-                }
-
-                m_ctx.cipher = nullptr;
-                m_ctx.cipher_data = nullptr;
-                m_ctx.engine = nullptr;
+#if OPENSSL_VERSION_NUMBER < 0x10100003L
+                EVP_CIPHER_CTX_cleanup(m_ctx);
+                m_ctx = nullptr;
```
Comment (singku) [inline aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp]: After openssl 1.1 m_ctx will be acquired from heap, so here we should free it right?

File: aws-cpp-sdk-core/include/aws/core/utils/crypto/openssl/CryptoImpl.h [hunk 5]
```diff
@@ -21,10 +21,16 @@
 #include <aws/core/utils/GetTheLights.h>
 #include <openssl/ossl_typ.h>
 #include <openssl/evp.h>
+#include <openssl/hmac.h>
 #include <openssl/rand.h>
 #include <atomic>
 #include <mutex>
 
+#if defined(LIBRESSL_VERSION_NUMBER) && (OPENSSL_VERSION_NUMBER == 0x20000000L)
```
Comment (andred) [inline aws-cpp-sdk-core/include/aws/core/utils/crypto/openssl/CryptoImpl.h]: I don't think it makes sense to speculate what LibreSSL might or might not do in the future. If they change the API, code using LibreSSL will have to follow. This is no different behaviour than before this PR - with the current code-base, if LibreSSL decided to change the API tomorrow, things would break.

The approach suggested here is what I've seen several other open-source projects taking.
It avoids additional clutter, as it redefines the version number to correctly reflect the API that LibreSSL actually implements.

The alternative would be to have implement proper configuration checks using CMake, to detect features / APIs, rather than clutter the code with ifdefs. I don't really care about LibreSSL, so if you want I can drop this commit. It looked like a low-hanging fruit.

File: aws-cpp-sdk-core/include/aws/core/utils/crypto/openssl/CryptoImpl.h (hunk 1 above)
Comment (andred) [inline aws-cpp-sdk-core/include/aws/core/utils/crypto/openssl/CryptoImpl.h]: As per the commit message:

> ...
> While LibreSSL strives to be API compatible with OpenSSL,
> it does so against the version when it was forked, 1.0.1g
> ...

File: aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp (hunk 3 above)
Comment (andred) [inline aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp]: Well spotted. EVP_CIPHER_CTX_cleanup() has been removed in OpenSSL 1.1 and instead EVP_CIPHER_CTX_reset() is to be used, so this adds another ifdef,

I'll do that.

File: aws-cpp-sdk-core/include/aws/core/utils/crypto/openssl/CryptoImpl.h (hunk 2 above)
Comment (andred) [inline aws-cpp-sdk-core/include/aws/core/utils/crypto/openssl/CryptoImpl.h]: I agree. Do you have a better suggestion? What about two subclasses, where each of them implements specifics for each OpenSSL version. I am not using C++ much day-to-day, so there might be better options?

**Comment by singku** [comment]
I have pulled this PR to our local branch, after reviewing and making proper changes, we will push out to this repo, probably in early next week.

File: aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp (hunk 4 above)
Comment (andred) [inline aws-cpp-sdk-core/source/utils/crypto/openssl/CryptoImpl.cpp]: I think this is OK. The code paths are slightly different (I think it's a bit ugly and hard to follow, actually). The idea was to be able to reuse the same EVP_CIPHER_CTX throughout the lifetime of a class OpenSSLCipher rather then freeing and re-allocating to achieve a more static memory allocation modelling, comparable to the OpenSSL 1.0 implementation.

Cleanup() is called from the destructor and from ::Reset(). For OpenSSL 1.1:
- Cleanup() itself doesn't do anything to the EVP_CIPHER_CTX
- the destructor does Cleanup(), followed by EVP_CIPHER_CTX_free()
- ::Reset() does Cleanup(), followed by Init(), which itself does EVP_CIPHER_CTX_reset()
hence we achieve re-use of the handle, rather than free() + malloc()

For OpenSSL 1.0:
- Cleanup() itself invalidates the static handle as per your comment
- the destructor does Cleanup()
- ::Reset() does Cleanup(), followed by Init(), which itself reinitialises the static handle

**Comment by andred** [comment]
> Why don't we just switch to pointer impl if OPENSSL made the type opaque?

OpenSSL 1.0 doesn't have APIs for allocation: https://github.com/aws/aws-sdk-cpp/pull/362#discussion_r111515671

**Comment by singku** [comment]
We have merged this pull request and did several changes. Check out the latest code in master branch.
